### PR TITLE
nixos/nvidia: assert `open` option is manually set on drivers >= 560

### PR DESCRIPTION
**Pull request by getchoo**
## Description of changes

This requirement was introduced in https://github.com/NixOS/nixpkgs/pull/337289 as a way to make sure users "explicitly pick which version of the driver they want since nvidia recommends the open one, but that is incompatible with older drivers". This is reasonable, however the user isn't informed in any real way aside from the upcoming release notes

This has caused a [good](https://github.com/NixOS/nixpkgs/pull/337289#issuecomment-2313802016) [amount](https://github.com/NixOS/nixpkgs/pull/337289#issuecomment-2318571311) [of](https://github.com/NixOS/nixpkgs/issues/338196) [confusion](https://github.com/NixOS/nixos-hardware/issues/1092) amongst users. By introducing this assertion and using a new `useOpenModules` local variable, we can have the same behavior but display a proper error message to hopefully clear things up until we can safely make this a default

https://github.com/NixOS/nixpkgs/pull/338362 was also mentioned as an alternative to this [here](https://github.com/NixOS/nixpkgs/pull/337289#discussion_r1739713732), however given our special case of it only needing to be set sometimes, I believe giving this message explicitly is a lot more user friendly

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/hardware/video/nvidia.nix [hunk 1]
```diff
@@ -287,6 +290,10 @@ in
               assertion = !(nvidiaEnabled && cfg.datacenter.enable);
               message = "You cannot configure both X11 and Data Center drivers at the same time.";
             }
+            {
+              assertion = cfg.open != null;
+              message = "`hardware.nvidia.open` must be set when using driver versions >= 560.";
```
Comment (Majiir) [inline nixos/modules/hardware/video/nvidia.nix:295]: "`hardware.nvidia.open` must be set" might be misinterpreted by some as "`hardware.nvidia.open` must be **`true`**".

**Comment by Kranzes** [comment]
I think there are some implications to this change that need to be considered. Nvidia does not recommend and does not support (?) GPUs prior to Turing with the open drivers. This change might cause the GPU to stop working for users with an older Nvidia GPU.

**Comment by getchoo** [comment]
> This change might cause the GPU to stop working for users with an older Nvidia GPU.

What part of this change exactly? No defaults are being set here; it's only explicitly documenting the requirement of setting this option, as introduced in the above PR. The only change will be the error message when this option isn't set

**Comment by Kranzes** [comment]
I've not been following too closely the history of this module but it seems this change now sets a default value to the `open` option. I think previously you had to pick a version, but with this change it will just default to `open` which may break systems with older GPUs.

**Comment by kanashimia** [comment]
> but with this change it will just default to `open` which may break systems with older GPUs.
No, there is an assertion that tests `cfg.open != null`

**Comment by kanashimia** [comment]
It could also fixed by allowing to define custom message for such options. https://github.com/NixOS/nixpkgs/pull/338362 improved message, but it is still bad because it isn't custom. @SuperSandro2000
Personally I think assertions are good enough.

File: nixos/modules/hardware/video/nvidia.nix [hunk 2]
```diff
@@ -287,6 +290,10 @@ in
               assertion = !(nvidiaEnabled && cfg.datacenter.enable);
               message = "You cannot configure both X11 and Data Center drivers at the same time.";
             }
+            {
+              assertion = cfg.open != null;
+              message = "`hardware.nvidia.open` must be defined when using driver versions >= 560.";
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/modules/hardware/video/nvidia.nix:295]: Let's use open source and closed source drivers to not allow any confusion here.

File: nixos/modules/hardware/video/nvidia.nix [hunk 3]
```diff
@@ -287,6 +290,10 @@ in
...
+            }
```
[the remark this answers is not in the document]
Comment (getchoo) [reply nixos/modules/hardware/video/nvidia.nix:296]: I believe the open modules support the datacenter drivers as well, so putting this behind the `nvidiaEnabled` conditional (which only checks if `services.xserver.videoDrivers` includes `"nvidia"`) with some of the other assertions could cause this to not work in the event of only `hardware.nvidia.datacenter.enable` being set

**Review by SuperSandro2000 — APPROVED** [review]
LGTM

File: nixos/modules/hardware/video/nvidia.nix [hunk 4]
```diff
@@ -257,15 +259,17 @@ in
...
+          if lib.versionOlder config.hardware.nvidia.package.version "560" then false else null
         '';
       };
 
       gsp.enable = lib.mkEnableOption ''
         the GPU System Processor (GSP) on the video card
       '' // {
+        default = useOpenModules || lib.versionAtLeast nvidia_x11.version "555";
         defaultText = lib.literalExpression ''
           config.hardware.nvidia.open || lib.versionAtLeast config.hardware.nvidia.package.version "555"
```
Comment (kanashimia) [inline nixos/modules/hardware/video/nvidia.nix:274]:
```suggestion
          config.hardware.nvidia.open == true || lib.versionAtLeast config.hardware.nvidia.package.version "555"
```

**Comment by kanashimia** [comment]
Looks good

**Review by kanashimia — APPROVED** [review]

**Pull request merged**